Fix comments query loop pagination not respecting Discussion Settings

tomasztunik opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Fixes so that Comments Query Loop block respects Discussion settings if comments pagination is disabled.

Fixes #39444

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Currently, if in the Discussion Settings paginating comments was not enabled PHP on the front-end would correctly render all comments and would not render the pagination. In the editor, these settings were not respected. Pagination was rendered and the set `per_page` value would be reflected in the number of rendered comments.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
On the editor side, I've added checking if paginating comments was enabled and if so I made the comments-pagination/editor not render and made sure that 

## Gotchas! 💣 💥 
- Custom query on the backend has no problem with fetching any number of comments. On the editor side though we were using WP REST endpoint to fetch comments and it has a hard limit of max 100 comments per page. We've decided after a conversation on #core-editor Slack that for the purpose of layout design and setup this limit should be enough and we shouldn't aggregate all pages to render all possible data.
- We shouldn't remove the pagination from the template if pagination was disabled at the time of adding the block as it wouldn't show up after the page/post was saved and comments pagination was enabled later. - _this was my first take..._
- In theory Discussion settings allow setting more than 100 items per page. This would fail in the Editor as if we tried to fetch from WP REST API more than 100 items per page it would throw an error but would work alright on the editor. - _added a guard for that_


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
1. Follow the steps from #39444
3. _With disabled comments pagination in the Discussion Settings_
3. The pagination should not show up and all comments (up to 100) should show up

cc: @c4rl0sbr4v0 @Mamaduka

tomasztunik committed:
hide comments pagination in the editor if not enabled

tomasztunik committed:
if comments pagination not enabled fetch max allowed items

+ fix possible error where if someone set more than 100 it would return
zero results in the editor but work ok when rendered by the backend

tomasztunik commented:
These CI failures don't seem to be related 

Noticed a 6 hours job timeout issue that creates $$$ and CO2 overhead. I've shared more on #core-editor [slack thread](https://wordpress.slack.com/archives/C02QB2JS7/p1649827656709219)

michalczaplinski reviewed (approved):
Thanks @tomasztunik ! Very nice PR 👏

Mamaduka commented on packages/block-library/src/comments-pagination/edit.js:
```diff
@@ -63,6 +64,21 @@ export default function QueryPaginationEdit( {
+	// controls. We don't want to remove them from the template so that when the user enables
+	// paging comments, the controls will be visible.
+	if ( ! pageComments ) {
+		return null;
```
Maybe we should explain why the editor doesn't display the actual pagination.

tomasztunik replied:
Added a comment 👍

Mamaduka replied:
Sorry for the confusion. I meant explaining to the user. Currently, this will render nothing, and folks might think it's a bug.

tomasztunik replied:
Ah thought it might be one of the two but this was quick enough that decided to see if it was the right call :)

Can you point me toward an example for similar conditional controls/UI elements? If we were to announce it to the user better - would it be better to just render the control but annotate it as "hidden" because of a setting, or render a warning/placeholder saying pagination would be here but it is disabled by settings?

Will roll back the comment for now.

Mamaduka replied:
The only thing that comes to mind is the Reusable block message that it cannot be rendered inside itself 😅  We can probably skip the warning wrapper.

https://github.com/WordPress/gutenberg/blob/0070d202cbb60407e3ec6d323dd7cb95150999ad/packages/block-library/src/block/edit.js#L83

Do you know who was leading design efforts for the Comment blocks? Maybe they have suggestions as well.

tomasztunik replied:
Hey! Actually, the `<Warning />` component looks pretty good I think?

![Screenshot 2022-04-14 at 15 46 52](https://user-images.githubusercontent.com/112270/163404313-442861b9-432b-44bb-af1d-20d2cb258f35.jpg)

Just how to word it properly? 

If we agree on a copy of this warning I can push it ⏩

Mamaduka replied:
Text looks good to me 👍

tomasztunik replied:
I've pushed the code but now I'm thinking... should "Discussion Settings" be a link? :)

Mamaduka replied:
Sorry, not sure about that.

michalczaplinski replied:
I'm not 100% convinced that we should add the `<Warning />` here if the lack of pagination is the expected and correct behaviour. I would lean towards not rendering any additional UI elements. 

I think it might be confusing if we use the same UI element when there is an actual error (e.g. when we show the `This block has encountered an error and cannot be previewed`) **and** also use this element to just inform the user why no pagination is being displayed.

@jameskoster was involved in the design of some of the Comment blocks so perhaps he'd like to chime in?

tomasztunik replied:
I've had a quick look and the `<Warning />` component seems to be used in a couple of other blocks including comments related blocks to inform about data availability and settings state. Ie. [PostCommentsForm](https://github.com/WordPress/gutenberg/pull/24162/files#diff-abf88ae09f89f6af7894b1f8480350eb820df74aff7e2b61490e93dd48c4722fR48-R52). The message though could be aligned with these as they all include the block name with a message following it.

I think there is value in the warning, and semantically it seems correct if we assume that the warning is not for the user who disabled the pagination but rather a warning to let other users (or the future "me") working with the blocks about why the pagination is not there as there is no obvious way to learn that without remembering to go to settings to review that.

waiting for more insights here :) 

(updated the copy to match how warnings are used in other comments related blocks)

michalczaplinski replied:
Yeah, that makes sense. I withdraw my objections since there is precedent for using the `<Warning/>` to inform rather than show errors as you point out. 🙂 

Re: the copy - I think it makes sense if we link to the Discussion settings. The only problem is that I don't see a reliable way to determine the admin URL. There was something custom added to that effect as a global [`window.wpAdminManifestL10n`](https://github.com/WordPress/gutenberg/blob/2a2938f254be29118b5dea5536d5444bc0076e6c/lib/pwa.php#L14-L15) which is then used [here](https://github.com/WordPress/gutenberg/blob/2a2938f254be29118b5dea5536d5444bc0076e6c/packages/admin-manifest/src/index.js#L114-L115) but it's an experimental package and we can't rely on it.

Mamaduka commented on packages/block-library/src/comments-pagination/edit.js:
```diff
@@ -63,6 +64,21 @@ export default function QueryPaginationEdit( {
 	} );
+
+	// Get the Discussion settings
+	const { pageComments } = useSelect( ( select ) => {
+		const { getSettings } = select( blockEditorStore );
+		const { __experimentalDiscussionSettings } = getSettings();
+		return __experimentalDiscussionSettings;
+	} );
```
```suggestion
	const pageComments = useSelect( ( select ) => {
		const { getSettings } = select( blockEditorStore );
		const { __experimentalDiscussionSettings } = getSettings();
		return __experimentalDiscussionSettings?.pageComments;
	}, [] );
```

I would return only the data we need in the component. I doubt `__experimentalDiscussionSettings` will be updated often, but it's just good practice :)

tomasztunik replied:
agreed, thanks for pointing it out and adding the dependencies array - missed it!

tomasztunik committed:
update selector to return only used data

Co-authored-by: George Mamadashvili <georgemamadashvili@gmail.com>

tomasztunik committed:
clarify that in editor we are rendering pagination placeholder

this is to enable styling the controls regardless of data and state

tomasztunik committed:
Present a user facing warning if comments pagination was disabled.

This will take place of the pagination placeholder.
+ Remove unneeded comment about comments pagination being a placeholder.

tomasztunik committed:
align disabled comments warning copy with other comments blocks

cbravobernal commented:
Thanks for the PR!

Are 100 results a lot to show in the editor? 

This could affect the performance and the editing experience. I guess there won't be many cases of users adding this out of the Site Editor (like in a post or page), but I guess that limiting the results to about 10 would be a good number.

tomasztunik commented:
Good point, but then if there are is more data should we show a warning
that results are limited in the editor?

pon., 18 kwi 2022, 10:36 użytkownik Carlos Bravo ***@***.***>
napisał:

> Thanks for the PR!
>
> Are 100 results a lot to show in the editor?
>
> This could affect the performance and the editing experience. I guess
> there won't be many cases of users adding this out of the Site Editor (like
> in a post or page), but I guess that limiting the results to about 10 would
> be a good number.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/40284#issuecomment-1101221047>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA3NDX2PKHCCXWY36A34GLVFUNJTANCNFSM5TIXCW7Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

cbravobernal commented:
> Good point, but then if there are is more data should we show a warning
that results are limited in the editor?

Another option could be to show that information on the block description. I'm not sure if always showing a warning is a good UX. Let's add a `needs design` label and [mention it on the Slack channel](https://wordpress.slack.com/archives/C02S78ZAL/p1650292972662369).

cbravobernal commented:
Answer from design contributors, [link to slack](https://wordpress.slack.com/archives/C02S78ZAL/p1650367477108569?thread_ts=1650292972.662369&cid=C02S78ZAL):

> Hey, good question, thanks for bringing it up.
In this case, there’s a similar issue with the regular (post) query loop, which also limits the amount shown in the editor (and even list view). As a baseline, I’d do the same for the comments query loop if you’re able to.
In both cases, we could add some help text indicating that the visual display is limited in the editor only, and why. But I’d add this help text to the inspector, potentially under the block description. I’d also do it as a separate followup PR, just since it hasn’t been problematic for the query loop.

Let's address it in another PR @tomasztunik , thanks for the contribution!

gziolo commented:
Cherry-picked to the `wp/6.0` branch for WordPress Beta 3 release.
